### ISSUE TITLE
Add a function for returning a byte string from a native string to help Python 3 porting

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6114]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#6114|
|Type|enhancement|
|Created|2012-10-16 12:37:22Z|
|Branch|https://github.com/twisted/twisted/tree/nativestring-py3-6114|

Many APIs only accept byte strings.  On Python 3, only native (unicode) strings support string interpolation though.  This makes it convenient to use native strings to construct a value (eg to be written to a socket), but Python 3 then requires an encode step whereas Python 2 does not.  A function to smooth over this difference would simplify some code.

<details><summary>Searchable metadata</summary>

```
trac-id__6114 6114
type__enhancement enhancement
reporter__exarkun exarkun
priority__normal normal
milestone__Python_3_3_Minimal Python 3.3 Minimal
branch__branches_nativestring_py3_6114 branches/nativestring-py3-6114
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1350391042000000 1350391042000000
changetime__1350404011000000 1350404011000000
version__None None
owner__exarkun exarkun

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [36178]) Branching to 'nativestring-py3-6114'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [36180]) Pull in code from http branch

refs [#6114](https://github.com/twisted/twisted/issues/6114)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @itamarst|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_4"></a>|@itamarst set owner to @exarkun|
|-|-|

  1. Add a `@param` indicating it takes a `str` as input.
  1. Given it has a particular use case, maybe the docstring should mention that, i.e. give an example of using it for string formatting?
  1. Add new function to `__all__`.

Fix, then merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

(In [36204]) Example, @param, and `__all__` update

refs [#6114](https://github.com/twisted/twisted/issues/6114)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set status to `closed`|
|-|-|

(In [36205]) Merge nativestring-py3-6114 (sic)

Author: exarkun
Reviewer: itamarst
Fixes: [#6114](https://github.com/twisted/twisted/issues/6114)

Introduce `twisted.python.compat.networkString`, a helper for formatting bytes on
Python 3.
